In [1]:
#import libraries
import pandas as pd
import numpy as np

In [2]:
#load data
data = pd.read_csv('gene_annot_test.tsv',sep='\t')
data

coords   outcome   score  length prev_annot  \
0          chr1:261635-267302:-1  accepted     888    5668        yes   
1          chr1:259026-261549:-1  accepted     650    2524        yes   
2          chr1:732208-739802:-1  rejected       0    7595         no   
3          chr1:720201-732016:-1  accepted       0   11816        yes   
4          chr1:711923-720031:-1  accepted      27    8109        yes   
...                          ...       ...     ...     ...        ...   
11005   chrY:20582694-20584473:1  accepted  278936    1780        yes   
11006   chrY:20584525-20588023:1  accepted  286043    3499        yes   
11007   chrY:20588106-20589483:1  accepted  444721    1378        yes   
11008   chrY:20589576-20592340:1  accepted  468983    2765        yes   
11009  chrY:20726098-20734468:-1  accepted       9    8371         no   

        transcript_source intron_sources splice_site  \
0                  SLRseq        SLR,CLS      GT..AG   
1                  SLRseq        SLR,CLS      GT..AG   
2      PacBio Capture-seq            CLS      GT..AG   
3      PacBio Capture-seq            CLS      GT..AG   
4      PacBio Capture-seq            CLS      GT..AG   
...                   ...            ...         ...   
11005  PacBio Capture-seq    SLR,CLS,RAC      GT..AG   
11006  PacBio Capture-seq    SLR,CLS,RAC      GT..AG   
11007  PacBio Capture-seq    SLR,CLS,RAC      GT..AG   
11008  PacBio Capture-seq    SLR,CLS,RAC      GT..AG   
11009  PacBio Capture-seq            CLS      GT..AG   

                                        repeat_overlap ss_antisense  \
0                              Type I Transposons/LINE           no   
1                              Type I Transposons/LINE           no   
2                              Type I Transposons/SINE           no   
3                                  Type II Transposons           no   
4                                           No overlap           no   
...                                                ...          ...   
11005                                       No overlap           no   
11006                                       No overlap           no   
11007                                       No overlap           no   
11008                                       No overlap           no   
11009  Type I Transposons/LINE,Type I Transposons/SINE           no   

        rej_reason annot_match incorrect_locus opp_strand false_ret_int  \
0              NaN         yes              no         no            no   
1              NaN         yes              no         no            no   
2              NaN          no              no         no            no   
3              NaN          no              no         no            no   
4              NaN          no              no         no            no   
...            ...         ...             ...        ...           ...   
11005  seq quality         yes              no         no            no   
11006  seq quality         yes              no         no            no   
11007  seq quality         yes              no         no            no   
11008  seq quality         yes              no         no            no   
11009          NaN          no              no         no            no   

            transcript_id                             gtype    bbiotype  \
0      OTTHUMT00000499557  transcribed_processed_pseudogene  non-coding   
1      OTTHUMT00000499557  transcribed_processed_pseudogene  non-coding   
2      OTTHUMT00000500170  transcribed_processed_pseudogene  non-coding   
3      OTTHUMT00000500170  transcribed_processed_pseudogene  non-coding   
4      OTTHUMT00000500170  transcribed_processed_pseudogene  non-coding   
...                   ...                               ...         ...   
11005  OTTHUMT00000500440                    protein_coding      coding   
11006  OTTHUMT00000500440                    protein_coding      coding   
11007  OTTHUMT00000500440                    pr

In [3]:
print(data.columns)
print(data.dtypes)

Index(['coords', 'outcome', 'score', 'length', 'prev_annot',
       'transcript_source', 'intron_sources', 'splice_site', 'repeat_overlap',
       'ss_antisense', 'rej_reason', 'annot_match', 'incorrect_locus',
       'opp_strand', 'false_ret_int', 'transcript_id', 'gtype', 'bbiotype',
       'rel_int_sup', 'rel_int_sup_k'],
      dtype='object')
coords                object
outcome               object
score                  int64
length                 int64
prev_annot            object
transcript_source     object
intron_sources        object
splice_site           object
repeat_overlap        object
ss_antisense          object
rej_reason            object
annot_match           object
incorrect_locus       object
opp_strand            object
false_ret_int         object
transcript_id         object
gtype                 object
bbiotype              object
rel_int_sup          float64
rel_int_sup_k        float64
dtype: object


In [4]:
#check for missing values
data.isna().sum()

coords                   0
outcome                  0
score                    0
length                   0
prev_annot               0
transcript_source        0
intron_sources           0
splice_site              0
repeat_overlap           0
ss_antisense             0
rej_reason           10315
annot_match              0
incorrect_locus          0
opp_strand               0
false_ret_int            0
transcript_id            0
gtype                    0
bbiotype                 0
rel_int_sup            783
rel_int_sup_k         2309
dtype: int64

In [5]:
# Drop the column due to 10315 missing values which is more than 75%
data = data.drop('rej_reason', axis=1)

In [6]:
# fill the missing values by taking mean of the rest of values
from sklearn.impute import SimpleImputer
imputer = SimpleImputer(missing_values = np.NaN,strategy = 'mean')
imputer = imputer.fit(data.iloc[:,17:20])
data.iloc[:,17:19] = imputer.transform(data.iloc[:,17:19])
data

coords   outcome   score  length prev_annot  \
0          chr1:261635-267302:-1  accepted     888    5668        yes   
1          chr1:259026-261549:-1  accepted     650    2524        yes   
2          chr1:732208-739802:-1  rejected       0    7595         no   
3          chr1:720201-732016:-1  accepted       0   11816        yes   
4          chr1:711923-720031:-1  accepted      27    8109        yes   
...                          ...       ...     ...     ...        ...   
11005   chrY:20582694-20584473:1  accepted  278936    1780        yes   
11006   chrY:20584525-20588023:1  accepted  286043    3499        yes   
11007   chrY:20588106-20589483:1  accepted  444721    1378        yes   
11008   chrY:20589576-20592340:1  accepted  468983    2765        yes   
11009  chrY:20726098-20734468:-1  accepted       9    8371         no   

        transcript_source intron_sources splice_site  \
0                  SLRseq        SLR,CLS      GT..AG   
1                  SLRseq        SLR,CLS      GT..AG   
2      PacBio Capture-seq            CLS      GT..AG   
3      PacBio Capture-seq            CLS      GT..AG   
4      PacBio Capture-seq            CLS      GT..AG   
...                   ...            ...         ...   
11005  PacBio Capture-seq    SLR,CLS,RAC      GT..AG   
11006  PacBio Capture-seq    SLR,CLS,RAC      GT..AG   
11007  PacBio Capture-seq    SLR,CLS,RAC      GT..AG   
11008  PacBio Capture-seq    SLR,CLS,RAC      GT..AG   
11009  PacBio Capture-seq            CLS      GT..AG   

                                        repeat_overlap ss_antisense  \
0                              Type I Transposons/LINE           no   
1                              Type I Transposons/LINE           no   
2                              Type I Transposons/SINE           no   
3                                  Type II Transposons           no   
4                                           No overlap           no   
...                                                ...          ...   
11005                                       No overlap           no   
11006                                       No overlap           no   
11007                                       No overlap           no   
11008                                       No overlap           no   
11009  Type I Transposons/LINE,Type I Transposons/SINE           no   

      annot_match incorrect_locus opp_strand false_ret_int  \
0             yes              no         no            no   
1             yes              no         no            no   
2              no              no         no            no   
3              no              no         no            no   
4              no              no         no            no   
...           ...             ...        ...           ...   
11005         yes              no         no            no   
11006         yes              no         no            no   
11007         yes              no         no            no   
11008         yes              no         no            no   
11009          no              no         no            no   

            transcript_id                             gtype    bbiotype  \
0      OTTHUMT00000499557  transcribed_processed_pseudogene  non-coding   
1      OTTHUMT00000499557  transcribed_processed_pseudogene  non-coding   
2      OTTHUMT00000500170  transcribed_processed_pseudogene  non-coding   
3      OTTHUMT00000500170  transcribed_processed_pseudogene  non-coding   
4      OTTHUMT00000500170  transcribed_processed_pseudogene  non-coding   
...                   ...                               ...         ...   
11005  OTTHUMT00000500440                    protein_coding      coding   
11006  OTTHUMT00000500440                    protein_coding      coding   
11007  OTTHUMT00000500440                    protein_coding      coding   
11008  OTTHUMT00000500440                    protein_coding      coding   
11009  OTTHUMT00000500305                         com

In [7]:
data.isna().sum()

coords               0
outcome              0
score                0
length               0
prev_annot           0
transcript_source    0
intron_sources       0
splice_site          0
repeat_overlap       0
ss_antisense         0
annot_match          0
incorrect_locus      0
opp_strand           0
false_ret_int        0
transcript_id        0
gtype                0
bbiotype             0
rel_int_sup          0
rel_int_sup_k        0
dtype: int64

In [8]:
# count the number of instances for each outcome label
outcome_counts = data['outcome'].value_counts()

# print the outcome counts
print(outcome_counts)

accepted    10601
rejected      409
Name: outcome, dtype: int64


In [9]:
# calculate the proportion of each outcome label
outcome_proportions = outcome_counts / len(data)

# print the outcome proportions
print(outcome_proportions)

accepted    0.962852
rejected    0.037148
Name: outcome, dtype: float64


**The accepted label has a significantly higher proportion than the rejected label, so the dataset is imbalanced, hence techniques such as data augmentation or oversampling should be done, doing undersampling will result in loss of vital info hence it is not recommended**

In [10]:
# split into independent and dependent varaible
X = data.iloc[:, 2:19]
y = data.iloc[:, 1]

In [11]:
X

score  length prev_annot   transcript_source intron_sources  \
0         888    5668        yes              SLRseq        SLR,CLS   
1         650    2524        yes              SLRseq        SLR,CLS   
2           0    7595         no  PacBio Capture-seq            CLS   
3           0   11816        yes  PacBio Capture-seq            CLS   
4          27    8109        yes  PacBio Capture-seq            CLS   
...       ...     ...        ...                 ...            ...   
11005  278936    1780        yes  PacBio Capture-seq    SLR,CLS,RAC   
11006  286043    3499        yes  PacBio Capture-seq    SLR,CLS,RAC   
11007  444721    1378        yes  PacBio Capture-seq    SLR,CLS,RAC   
11008  468983    2765        yes  PacBio Capture-seq    SLR,CLS,RAC   
11009       9    8371         no  PacBio Capture-seq            CLS   

      splice_site                                   repeat_overlap  \
0          GT..AG                          Type I Transposons/LINE   
1          GT..AG                          Type I Transposons/LINE   
2          GT..AG                          Type I Transposons/SINE   
3          GT..AG                              Type II Transposons   
4          GT..AG                                       No overlap   
...           ...                                              ...   
11005      GT..AG                                       No overlap   
11006      GT..AG                                       No overlap   
11007      GT..AG                                       No overlap   
11008      GT..AG                                       No overlap   
11009      GT..AG  Type I Transposons/LINE,Type I Transposons/SINE   

      ss_antisense annot_match incorrect_locus opp_strand false_ret_int  \
0               no         yes              no         no            no   
1               no         yes              no         no            no   
2               no          no              no         no            no   
3               no          no              no         no            no   
4               no          no              no         no            no   
...            ...         ...             ...        ...           ...   
11005           no         yes              no         no            no   
11006           no         yes              no         no            no   
11007           no         yes              no         no            no   
11008           no         yes              no         no            no   
11009           no          no              no         no            no   

            transcript_id                             gtype    bbiotype  \
0      OTTHUMT00000499557  transcribed_processed_pseudogene  non-coding   
1      OTTHUMT00000499557  transcribed_processed_pseudogene  non-coding   
2      OTTHUMT00000500170  transcribed_processed_pseudogene  non-coding   
3      OTTHUMT00000500170  transcribed_processed_pseudogene  non-coding   
4      OTTHUMT00000500170  transcribed_processed_pseudogene  non-coding   
...                   ...                               ...         ...   
11005  OTTHUMT00000500440                    protein_coding      coding   
11006  OTTHUMT00000500440                    protein_coding      coding   
11007  OTTHUMT00000500440                    protein_coding      coding   
11008  OTTHUMT00000500440                    protein_coding      coding   
11009  OTTHUMT00000500305                         comp_pipe       novel   

       rel_int_sup  rel_int_sup_k  
0         0.311999       0.311999  
1        -0.311999      -0.311999  
2        -2.525729      -2.708050  
3        -2.525729      -2.931194  
4         1.216395       0.810930  
...            ...            ...  
11005    -0.141738      -0.364851  
11006    -0.112146      -0.335259  
11007     0.433606       0.210496  
11008     0.503702       0.280593  
11009    -0.874467      -1.139485  

[11010 rows x 17 columns]

In [12]:
# Feature Scaling (0-1) on the numerical values otherwise they have the highest effect
from sklearn.preprocessing import StandardScaler
sc = StandardScaler()
X.iloc[:,[0,1,15,16]]= sc.fit_transform(X.iloc[:,[0,1,15,16]])
X

score    length prev_annot   transcript_source intron_sources  \
0     -0.257487 -0.145344        yes              SLRseq        SLR,CLS   
1     -0.257747 -0.274457        yes              SLRseq        SLR,CLS   
2     -0.258457 -0.066210         no  PacBio Capture-seq            CLS   
3     -0.258457  0.107131        yes  PacBio Capture-seq            CLS   
4     -0.258428 -0.045102        yes  PacBio Capture-seq            CLS   
...         ...       ...        ...                 ...            ...   
11005  0.046257 -0.305010        yes  PacBio Capture-seq    SLR,CLS,RAC   
11006  0.054021 -0.234417        yes  PacBio Capture-seq    SLR,CLS,RAC   
11007  0.227363 -0.321519        yes  PacBio Capture-seq    SLR,CLS,RAC   
11008  0.253868 -0.264560        yes  PacBio Capture-seq    SLR,CLS,RAC   
11009 -0.258447 -0.034342         no  PacBio Capture-seq            CLS   

      splice_site                                   repeat_overlap  \
0          GT..AG                          Type I Transposons/LINE   
1          GT..AG                          Type I Transposons/LINE   
2          GT..AG                          Type I Transposons/SINE   
3          GT..AG                              Type II Transposons   
4          GT..AG                                       No overlap   
...           ...                                              ...   
11005      GT..AG                                       No overlap   
11006      GT..AG                                       No overlap   
11007      GT..AG                                       No overlap   
11008      GT..AG                                       No overlap   
11009      GT..AG  Type I Transposons/LINE,Type I Transposons/SINE   

      ss_antisense annot_match incorrect_locus opp_strand false_ret_int  \
0               no         yes              no         no            no   
1               no         yes              no         no            no   
2               no          no              no         no            no   
3               no          no              no         no            no   
4               no          no              no         no            no   
...            ...         ...             ...        ...           ...   
11005           no         yes              no         no            no   
11006           no         yes              no         no            no   
11007           no         yes              no         no            no   
11008           no         yes              no         no            no   
11009           no          no              no         no            no   

            transcript_id                             gtype    bbiotype  \
0      OTTHUMT00000499557  transcribed_processed_pseudogene  non-coding   
1      OTTHUMT00000499557  transcribed_processed_pseudogene  non-coding   
2      OTTHUMT00000500170  transcribed_processed_pseudogene  non-coding   
3      OTTHUMT00000500170  transcribed_processed_pseudogene  non-coding   
4      OTTHUMT00000500170  transcribed_processed_pseudogene  non-coding   
...                   ...                               ...         ...   
11005  OTTHUMT00000500440                    protein_coding      coding   
11006  OTTHUMT00000500440                    protein_coding      coding   
11007  OTTHUMT00000500440                    protein_coding      coding   
11008  OTTHUMT00000500440                    protein_coding      coding   
11009  OTTHUMT00000500305                         comp_pipe       novel   

        rel_int_sup  rel_int_sup_k  
0      5.006334e-01       0.651294  
1      2.373351e-01       0.371300  
2     -6.967550e-01      -0.703829  
3     -6.967550e-01      -0.803955  
4      8.822461e-01       0.875169  
...             ...            ...  
11005  3.091776e-01       0.347585  
11006  3.216640e-01       0.360864  
11007  5.519456e-01       0.605749  
11008  5.815230e-01       0.637202  
11009  4.684621e-17       0.000000  

[11010 rows x 17 colu

In [13]:
X = X.drop('repeat_overlap', axis=1)

In [14]:
X = X.drop('transcript_id', axis=1)

In [15]:
X

score    length prev_annot   transcript_source intron_sources  \
0     -0.257487 -0.145344        yes              SLRseq        SLR,CLS   
1     -0.257747 -0.274457        yes              SLRseq        SLR,CLS   
2     -0.258457 -0.066210         no  PacBio Capture-seq            CLS   
3     -0.258457  0.107131        yes  PacBio Capture-seq            CLS   
4     -0.258428 -0.045102        yes  PacBio Capture-seq            CLS   
...         ...       ...        ...                 ...            ...   
11005  0.046257 -0.305010        yes  PacBio Capture-seq    SLR,CLS,RAC   
11006  0.054021 -0.234417        yes  PacBio Capture-seq    SLR,CLS,RAC   
11007  0.227363 -0.321519        yes  PacBio Capture-seq    SLR,CLS,RAC   
11008  0.253868 -0.264560        yes  PacBio Capture-seq    SLR,CLS,RAC   
11009 -0.258447 -0.034342         no  PacBio Capture-seq            CLS   

      splice_site ss_antisense annot_match incorrect_locus opp_strand  \
0          GT..AG           no         yes              no         no   
1          GT..AG           no         yes              no         no   
2          GT..AG           no          no              no         no   
3          GT..AG           no          no              no         no   
4          GT..AG           no          no              no         no   
...           ...          ...         ...             ...        ...   
11005      GT..AG           no         yes              no         no   
11006      GT..AG           no         yes              no         no   
11007      GT..AG           no         yes              no         no   
11008      GT..AG           no         yes              no         no   
11009      GT..AG           no          no              no         no   

      false_ret_int                             gtype    bbiotype  \
0                no  transcribed_processed_pseudogene  non-coding   
1                no  transcribed_processed_pseudogene  non-coding   
2                no  transcribed_processed_pseudogene  non-coding   
3                no  transcribed_processed_pseudogene  non-coding   
4                no  transcribed_processed_pseudogene  non-coding   
...             ...                               ...         ...   
11005            no                    protein_coding      coding   
11006            no                    protein_coding      coding   
11007            no                    protein_coding      coding   
11008            no                    protein_coding      coding   
11009            no                         comp_pipe       novel   

        rel_int_sup  rel_int_sup_k  
0      5.006334e-01       0.651294  
1      2.373351e-01       0.371300  
2     -6.967550e-01      -0.703829  
3     -6.967550e-01      -0.803955  
4      8.822461e-01       0.875169  
...             ...            ...  
11005  3.091776e-01       0.347585  
11006  3.216640e-01       0.360864  
11007  5.519456e-01       0.605749  
11008  5.815230e-01       0.637202  
11009  4.684621e-17       0.000000  

[11010 rows x 15 columns]

In [16]:
from sklearn.preprocessing import OneHotEncoder, LabelEncoder
from sklearn.compose import ColumnTransformer

In [17]:
# Specify the columns to be one-hot encoded
columns_to_encode = [2,3,4,5,6,7,8,9,10,11,12]  # indices of columns 

# Create the ColumnTransformer object
ct = ColumnTransformer([('encoder', OneHotEncoder(), columns_to_encode)], remainder='passthrough')

# Apply the transformer to the dataset
X = ct.fit_transform(X)
X[0]

array([ 0.        ,  1.        ,  0.        ,  0.        ,  0.        ,
        1.        ,  0.        ,  0.        ,  0.        ,  0.        ,
        1.        ,  0.        ,  0.        ,  0.        ,  0.        ,
        1.        ,  0.        ,  1.        ,  0.        ,  0.        ,
        1.        ,  1.        ,  0.        ,  1.        ,  0.        ,
        1.        ,  0.        ,  0.        ,  0.        ,  0.        ,
        0.        ,  0.        ,  0.        ,  1.        ,  0.        ,
        0.        ,  0.        ,  0.        ,  1.        ,  0.        ,
       -0.25748711, -0.14534438,  0.50063338,  0.65129391])

In [18]:
# 0 - accepted, 1 - rejected
labelencoder_y = LabelEncoder()
y = labelencoder_y.fit_transform(y)
y

array([0, 0, 1, ..., 0, 0, 0])

In [19]:
# split the data into train and test set
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.25, random_state = 0)

In [20]:
from imblearn.over_sampling import RandomOverSampler

In [21]:
os = RandomOverSampler()

In [22]:
# oversampling the data to balance imbalace data
X_train_res, y_train_res = os.fit_resample(X_train,y_train)

In [23]:
from collections import Counter

In [24]:
# counting labels of 0 and 1
print('Original Dataset shape {}'.format(Counter(y_train)))
print('Resampled Dataset shape {}'.format(Counter(y_train_res)))

Original Dataset shape Counter({0: 7958, 1: 299})
Resampled Dataset shape Counter({0: 7958, 1: 7958})


## Random Forest

In [25]:
from sklearn.ensemble import RandomForestClassifier
classifier = RandomForestClassifier()
classifier.fit(X_train_res,y_train_res)

RandomForestClassifier()

In [26]:
from sklearn.metrics import confusion_matrix, accuracy_score, classification_report

In [27]:
y_pred=classifier.predict(X_test)
print(confusion_matrix(y_test,y_pred))
print(accuracy_score(y_test,y_pred))
print(classification_report(y_test,y_pred))

[[2617   26]
 [  34   76]]
0.9782055938975663
              precision    recall  f1-score   support

           0       0.99      0.99      0.99      2643
           1       0.75      0.69      0.72       110

    accuracy                           0.98      2753
   macro avg       0.87      0.84      0.85      2753
weighted avg       0.98      0.98      0.98      2753



## Support Vector Classifier

In [28]:
from sklearn.svm import SVC
classifier = SVC(kernel = 'linear', random_state = 0)
classifier.fit(X_train, y_train)

SVC(kernel='linear', random_state=0)

In [29]:
y_pred=classifier.predict(X_test)
print(confusion_matrix(y_test,y_pred))
print(accuracy_score(y_test,y_pred))
print(classification_report(y_test,y_pred))

[[2625   18]
 [  47   63]]
0.9763893933890302
              precision    recall  f1-score   support

           0       0.98      0.99      0.99      2643
           1       0.78      0.57      0.66       110

    accuracy                           0.98      2753
   macro avg       0.88      0.78      0.82      2753
weighted avg       0.97      0.98      0.97      2753



## Decision Tree Classifier

In [30]:
from sklearn.tree import DecisionTreeClassifier
classifier = DecisionTreeClassifier(criterion='entropy', random_state=0)
classifier.fit(X_train, y_train)

DecisionTreeClassifier(criterion='entropy', random_state=0)

In [31]:
y_pred=classifier.predict(X_test)
print(confusion_matrix(y_test,y_pred))
print(accuracy_score(y_test,y_pred))
print(classification_report(y_test,y_pred))

[[2608   35]
 [  45   65]]
0.9709407918634217
              precision    recall  f1-score   support

           0       0.98      0.99      0.98      2643
           1       0.65      0.59      0.62       110

    accuracy                           0.97      2753
   macro avg       0.82      0.79      0.80      2753
weighted avg       0.97      0.97      0.97      2753



## From the above classifier(Random Forest, SVC, DT) we can see that all the supervised learning have accuracy > 97% which is good and also the data is labeled, so Supervised Learning is recommended
## Also Random Forest > SVC > DT acc to the decresing order of accuracy